seed/src/handle.rs

```diff
@@ -14,59 +19,71 @@ use crate::Project;
 #[derive(Debug, Error)]
 #[non_exhaustive]
 pub enum NodeError {
+    #[error("request response failed")]
+    RequestResponseFailed(#[from] oneshot::error::RecvError),
+
+    #[error("request send failed")]
+    RequestSendFailed,
+
     #[error("request failed: the node disconnected")]
     RequestFailed,
+
+    #[error("request timed out")]
+    RequestTimeout(#[from] time::error::Elapsed),
 }
 
 /// Handle used to interact with the seed node.
 pub struct NodeHandle {
-    channel: chan::UnboundedSender<Request>,
+    channel: mpsc::Sender<Request>,
 }
 
 impl NodeHandle {
-    pub(crate) fn new(channel: chan::UnboundedSender<Request>) -> Self {
+    pub(crate) fn new(channel: mpsc::Sender<Request>) -> Self {
         Self { channel }
     }
 
     pub async fn get_membership(&mut self) -> Result<MembershipInfo, NodeError> {
-        let (tx, mut rx) = chan::channel(1);
+        let (tx, rx) = oneshot::channel();
         self.channel
-            .send(Request::GetMembership(tx))
-            .await
-            .map_err(|_| NodeError::RequestFailed)?;
+            .try_send(Request::GetMembership(tx))
+            .map_err(|_| NodeError::RequestSendFailed)?;
 
-        rx.next().await.ok_or(NodeError::RequestFailed)
+        time::timeout(Duration::from_secs(3), rx)
```

Comment:
Shouldn't the caller decide on the timeout?

Reply:
That's an interesting alternative, it would complicate the API and I'd assume we want a default so we would need to expose an `Option` param or have a way for the NodeHandle to be configured. I'd be curious how the caller decides what the right value is. 3 is a pretty solid number.

Reply:
Yeah, either it's set on the handle, eg `handle.set_timeout(x)`, or you have two functions, one which takes a timeout, eg. `foo` and `foo_timeout`, or also just force the caller to specify a timeout, since it's a low level library function..

Just throwing some other options, though 3 is a good number.

Reply:
The thing I don't like about using an `Option` is when you read `handle.get_memberships(None)` it's like wut.